### DISCUSSION TITLE
Package info from meta.yaml not updated on anaconda.org

### DISCUSSION THREAD

**Discussion by lindstro** [conda/infra#649] [General]
I have a question about getting package metadata updated on anaconda.org.  Our [zfp/zfpy packages](https://github.com/conda-forge/zfpy-feedstock) were updated recently and contain the following in meta.yaml:
```
about:
  home: https://zfp.llnl.gov
  license: BSD-3-Clause
  license_family: BSD
  license_file: LICENSE
  summary: Library for compressed numerical arrays that support high throughput read and write random access
  description: |
    zfp is an open source C/C++ library for compressed numerical arrays
    that support high throughput read and write random access. zfp also
    supports streaming compression of integer and floating-point data,
    e.g., for applications that read and write large data sets to and
    from disk. zfp is primarily written in C and C++ but also includes
    Python and Fortran bindings.
  doc_url: https://zfp.readthedocs.io/
  dev_url: https://github.com/LLNL/zfp
```
The [anaconda package website](https://anaconda.org/conda-forge/zfp) does not reflect the updated `home`, `license`, or `doc_url` fields, however:

<img width="972" alt="image" src="https://user-images.githubusercontent.com/1304791/201448759-4efb208a-e5ef-470a-8940-098750af31b8.png">

The Files tab shows the package uploaded a couple of weeks ago.  I'd appreciate any suggestions on how to get the metadata updated.

**Comment by yshmatov-anaconda** [comment]
You should be able to update meta data of existing release with `anaconda update` command (_available since anaconda-client 1.9.0_).

To do this, create JSON file with your metadata, e.g.:
```
{
  "home": "https://zfp.llnl.gov",
  "license": "BSD-3-Clause",
  "license_family": "BSD",
  "license_file": "LICENSE",
  "summary": "Library for compressed numerical arrays that support high throughput read and write random access",
  "description": "zfp is an open source C/C++ library for compressed numerical arrays that support high throughput read and write random access. zfp also supports streaming compression of integer and floating-point data, e.g., for applications that read and write large data sets to and from disk. zfp is primarily written in C and C++ but also includes Python and Fortran bindings.",
  "doc_url": "https://zfp.readthedocs.io/",
  "dev_url": "https://github.com/LLNL/zfp"
}
```

And run `anaconda update` command:
```
anaconda update --release conda-forge/zfp/1.0.0 new_meta.json
```
Where last argument is the path to your JSON file with new metadata.

**Reply by jakirkham** [reply]
The issue is conda-forge uploads packages to a staging channel for validation and then [copies them over]( https://github.com/conda-forge/conda-forge-webservices/blob/ac84983eb66239c8d3bd6f5fb8b3297f709d2f8d/conda_forge_webservices/feedstock_outputs.py#L126 ). The metadata is lost during copying.

At one point thought this was fixed ( https://github.com/conda-forge/cudatoolkit-feedstock/issues/2#issuecomment-768463577 ). So must admit am a little surprised we are seeing this issue again.

cc @chenghlee @beckermr

**Reply by beckermr** [reply]
Indeed. This appears to be a regression.

**Reply by leofang** [reply]
This regression happened a while ago and impacted a number of our packages on conda-forge:
- https://github.com/conda-forge/cusparselt-feedstock/issues/14
- https://github.com/conda-forge/cuquantum-feedstock/issues/3
- https://github.com/conda-forge/cuda-python-feedstock/issues/7

It'd be nice to get them fixed asap.

**Reply by leofang** [reply]
The package info is also not updated in our new packages (license missing, even after we fixed the sub package about section)
- https://anaconda.org/conda-forge/custatevec
- https://anaconda.org/conda-forge/cutensornet

**Reply by jakirkham** [reply]
cc @jezdez

**Reply by jessewiles** [reply]
I'm able to confirm the issue in the packages you mention, but I'm having trouble replicating the behavior. Does the staging channel UI display the updated meta info on the initial upload? Also, is the new `meta.yaml` upload overwriting an existing package version in the staging channel or is it a part of a new package version?

**Reply by beckermr** [reply]
Check the api operation where copies from one channel are made to another.

**Reply by leofang** [reply]
@jessewiles This is my [rule of thumb](https://github.com/conda-forge/cuquantum-feedstock/issues/3#issuecomment-1006101949) based on observations. I could be wrong but so far it seems I am right 🙂 If the package info (license, license url, etc) is correctly configured at the **very first upload** of the package to anaconda.org, regardless if it's staging or conda-forge (i.e. it's uploaded  or copied), the info would be correct. Otherwise, the info would be incorrect.

**Reply by leofang** [reply]
> Otherwise, the info would be incorrect.

And cannot be fixed by any subsequent package update.

**Reply by jessewiles** [reply]
The `--replace` and `--update` flags on the `copy` anaconda client command can be used to trigger requests that result in metadata change.  The `--replace` flag causes the client to issue a `PUT` request and `--update` cause the client to issue a `PATCH` request.

Also of note, it's possible to update the metadata using the API directly.  For example, I used the following request to update the metadata [in a copy of zfp I created](https://anaconda.org/jwixel/zfp):

```
curl -X 'PATCH' \ 
        'https://api.anaconda.org/release/jwixel/zfp/1.0.0' \
        -H 'accept: application/json' \
        -H 'Content-Type: application/json' \
        -H 'Authorization: token <REDACT-TOKEN> \
        -d '{"public_attrs": {"doc_url": "https://zfp.readthedocs.io/", "dev_url": "https://github.com/LLNL/zfp", "home": "https://zfp.llnl.gov"}}'
```

**Reply by beckermr** [reply]
Ahhhh. I think we need to set `replace=True` here then: https://github.com/conda-forge/conda-forge-webservices/blob/main/conda_forge_webservices/feedstock_outputs.py#L126

**Reply by jessewiles** [reply]
> And cannot be fixed by any subsequent package update.

As @yshmatov-anaconda (and I) noted, the info can be fixed with an API request that is independent of the package upload or copy.

**Comment by beckermr** [comment]
Nope. I tried multiple iterations of uploading, deleting etc. Here is how to reproduce the bug:

1. Copy a new package to your channel. I used 

```
ac.copy(
    "conda-forge",
    "zfp",
    "0.5.5",
    basename=urllib.parse.quote("linux-64/zfp-0.5.5-he1b5a44_1.tar.bz2", safe=""),
    to_owner="beckermr",
    update=False,
    replace=False,
)
```

2. Delete this package from your personal channel. I did this in the UI.

3. Then copy a newer version

```
ac.copy(
    "conda-forge",
    "zfp",
    "1.0.0",
    basename=urllib.parse.quote("osx-64/zfp-1.0.0-ha894c9a_3.tar.bz2", safe=""),
    to_owner="beckermr",
    update=False,
    replace=True,
)
```

This new version has in its about.json

```
"home": "https://zfp.llnl.gov",
  "identifiers": [],
  "keywords": [],
  "license": "BSD-3-Clause",
  "license_family": "BSD",
  "license_file": "LICENSE",
```

and yet the UI still shows the info for version 0.5.5


<img width="1059" alt="Screenshot 2022-12-23 at 11 37 28 PM" src="https://user-images.githubusercontent.com/5296416/209423071-ca9e60c4-ed53-4a7b-a422-da5b20c9fd7d.png">

I tried with and without the `replace=True` keyword and it doesn't change anything. I tried multiple `replace=True` calls and that didn't do anything either.

As @leofang eluded to, it appears the info in the UI is only set once on the first upload and can be added to but not replaced.

**Reply by jessewiles** [reply]
> and yet the UI still shows the info for version 0.5.5

The UI is showing the info from the 1.0.0 database record at anaconda.org. I am not sure why/how the database record was not properly updated when the original 1.0.0 release was uploaded on cf-staging. This seems likely a bug and I'm working to replicate it.

I can verify both @beckermr's and @leofang's assertions regarding when the UI is updated. Following is a summary of how things work today.

The server API endpoints do not crack open the tarball and scan metadata information. All reading of metadata is handled by the client and then POSTed to the [/package/<owner_login>/<package_name>](https://github.com/Anaconda-Platform/anaconda-client/blob/2893410fa23ccd14e514c945a1f0e744adc7cb83/binstar_client/__init__.py#L357) or [/release/<owner_login>/<package_name>/\<version>](https://api.anaconda.org/docs#/package/post_release__owner_login___package_name___version_) API endpoints. Reading of metadata from the tarball and passing it to the server only occurs on [upload](https://github.com/Anaconda-Platform/anaconda-client/blob/2893410fa23ccd14e514c945a1f0e744adc7cb83/binstar_client/commands/upload.py#L185-L258), not on [copy](https://github.com/Anaconda-Platform/anaconda-client/blob/2893410fa23ccd14e514c945a1f0e744adc7cb83/binstar_client/mixins/package.py#L12-L33).

There are two places in the `upload_package` code where the metadata that the client reads from the package can be passed along to the server: on [add_package](https://github.com/Anaconda-Platform/anaconda-client/blob/2893410fa23ccd14e514c945a1f0e744adc7cb83/binstar_client/commands/upload.py#L215) and [add_release](https://github.com/Anaconda-Platform/anaconda-client/blob/2893410fa23ccd14e514c945a1f0e744adc7cb83/binstar_client/__init__.py#L455). If the package already exists, then [nothing happens in the add_package call](https://github.com/Anaconda-Platform/anaconda-client/blob/2893410fa23ccd14e514c945a1f0e744adc7cb83/binstar_client/commands/upload.py#L120).  Otherwise a new package is created including the metadata scanned from the metadata file within the tarball, which is how the initial metadata appears on a package.

The [add_release call](https://github.com/Anaconda-Platform/anaconda-client/blob/2893410fa23ccd14e514c945a1f0e744adc7cb83/binstar_client/commands/upload.py#L156-L164) results in a client request that [includes the release attribute metadata](https://github.com/Anaconda-Platform/anaconda-client/blob/2893410fa23ccd14e514c945a1f0e744adc7cb83/binstar_client/__init__.py#L465). The release metadata includes [all of the relevant KV pairs](https://github.com/Anaconda-Platform/anaconda-client/blob/2893410fa23ccd14e514c945a1f0e744adc7cb83/binstar_client/inspect_package/conda.py#L118-L132) from the `about` section of the metadata file.

The `--replace` and `--update` flags in the **copy** operation do result [in the `PUT/PATCH` requests](https://github.com/Anaconda-Platform/anaconda-client/blob/2893410fa23ccd14e514c945a1f0e744adc7cb83/binstar_client/mixins/package.py#L21-L24) as previously described. The PUT or PATCH operations are related to how the server updates the database record which contains the metadata information.  To clarify, the destination database record may either be updated (PATCHed) to match the source database record, or the source database record may replace (PUT) the destination database record.

Because the **copy** operation is handled entirely on the server and only triggered by the client, information within the release tarball is not scanned or used by the **copy** operation.

**Reply by leofang** [reply]
Hi @jessewiles Thank you for your clear explanation for the bug and for verifying our observation, and sorry that I dropped the ball. Do you have an ETA for the fix? Can we do anything to help you, e.g. verify the fix is working correctly?

**Reply by jessewiles** [reply]
Hi @leofang. Thanks for staying on this. We're still working through the best way to handle this case. I should have more information later this week.

**Reply by jessewiles** [reply]
Hi @leofang 👋 Would you be comfortable patching **anaconda-client** locally to test a proposed solution?

**Reply by jakirkham** [reply]
@jessewiles, think this gets into a deeper question. Since this would need to happen in conda-forge infra

What is the patch? Is there a PR up?

Maybe we can take a look and evaluate

**Reply by jessewiles** [reply]
@jakirkham Here is the [PR](https://github.com/Anaconda-Platform/anaconda-client/pull/641). To clarify, this PR allows for correcting the root of the problem. There are no plans to change how `copy` works as it would be a significant architectural shift from the way things work currently. To correct the issue with the `zfp` package, re-upload the package with the new `--use-package-metadata` flag. Subsequent copies should correctly reflect the updated metadata.

**Reply by leofang** [reply]
@beckermr Would it be possible for you to assess Jesse's patch for CF? Do we need to change any bot code, or just conda-smithy?

**Reply by jakirkham** [reply]
Thanks Jesse! 🙏

Leo, I think this just amounts to patch `anaconda-client` ( https://github.com/conda-forge/anaconda-client-feedstock/pull/38 ) and possibly updating the webservice responsible for uploading

**Reply by leofang** [reply]
So where does the web service code live? 😅

**Reply by beckermr** [reply]
The web service uses a copy and does not reupload. Reuploading is a security issue for more than one reason and would cause load issues on the webservices. We can merge the patch but someone will have to reupload by hand to fix things.

**Reply by leofang** [reply]
@beckermr After merging the patch, we need to modify the web service code to add the `--use-package-metadata` flag, right? And then we rebuild the affected packages to trigger the metadata update?

**Reply by jessewiles** [reply]
Heads up: we [ultimately landed on `force-metadata-update`](https://github.com/Anaconda-Platform/anaconda-client/pull/641) for the name of this option.

**Reply by jakirkham** [reply]
@jessewiles does this work when copying a package from one channel to another? This is the conda-forge use case as we upload packages to [`cf-staging`]( https://anaconda.org/cf-staging ) and then [copy them]( https://github.com/conda-forge/conda-forge-webservices/blob/3918b383c0a1a51d3d18b13b5b75770b4a6acb9f/conda_forge_webservices/feedstock_outputs.py#L142-L150 ) (via [`copy`]( https://github.com/Anaconda-Platform/anaconda-client/blob/bba35b31d2801ae3edf09ae35a130c85b02fe612/binstar_client/mixins/package.py#L15-L16 )) to [`conda-forge`]( https://anaconda.org/conda-forge )

**Reply by jessewiles** [reply]
@jakirkham No. The issue in this case is that the item that is being copied (_the one in cf-staging_) is also incorrect. The **copy** operation works entirely on the server. It will only propagate the existing data. The server does not inspect the packages. Stated a different way, the only way to affect the metadata which appears on the website is [by issuing a PATCH request](https://github.com/conda/infrastructure/discussions/649#discussioncomment-4489632), or by uploading the package again with the new `--force-metadata-update` flag.

To further clarify, this issue occurs whenever a new distribution is uploaded to an existing release (version, ie 1.0.0). In this case, zfp 1.0.0 was initially uploaded before the changes to the metadata. Later additional 1.0.0 distributions were uploaded with updated metadata. However the existing client code doesn't inspect the package metadata for a new distribution of an existing release. Once the client detects that a release (version) already exists on the server, the client skips ahead to the stage/upload/commit portion of the upload process.

The [new changes to the client](https://github.com/Anaconda-Platform/anaconda-client/pull/641) empower the user to force the upload to update the server metadata to match the metadata from the local package. HTH.

**Reply by jakirkham** [reply]
Gotcha thanks for the explanation Jesse! 🙏

So IIUC (and please correct me if not) `--force-metadata-update` flag when added would ensure the right metadata ends up on `cf-staging` (where the package is uploaded). When we perform the copy from `cf-staging` to `conda-forge`, all of the metadata should copy over. Is that correct or am I still missing something?

**Reply by jessewiles** [reply]
Yep, that's correct 👍

**Reply by leofang** [reply]
Thanks for clarifying @jessewiles @jakirkham. Sounds like as long as CF uploads to cf-staging with the new flag always on, we can ensure we keep the metadata up-to-date (even after copied from cf-staging to conda-forge).

**Reply by jakirkham** [reply]
Trying to see if PR ( https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/241 ) would move us in the right direction

**Reply by leofang** [reply]
> Trying to see if PR ( [conda-forge/conda-forge-ci-setup-feedstock#241](https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/241) ) would move us in the right direction

Hi @jessewiles and all, we've merged the above PR and [rebuilt the packages from the cuquantum-feedstock](https://github.com/conda-forge/cuquantum-feedstock/pull/39), which are troubled with missing license info on Anaconda.org. It appears that the new packages uploaded to [the cf-staging channel are fixed](https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/241#issuecomment-1589551667), but [not those copied over](https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/241#issuecomment-1589517475) from the cf-staging channel to the conda-forge channel. Something is still missing here.

**Reply by jessewiles** [reply]
@leofang The license information for the conda-forge record is correct in the database (ie, it matches the cf-staging record). I think this is a caching problem. Let me verify.

**Reply by jessewiles** [reply]
OK, more context. On the backend there is a distinction between the **release** and the **package**. The **release** record is correct. However the metadata portion of the web page is rendered from the **package**, which still has the old data. This is the purpose of the `--replace` flag in binstar-client's `copy` command. Using `--replace` with the copy will trigger a `PUT` which will cause the backend server to update the _target_ **package**'s database record with the details of the _source's_ database record (_sans id and timestamps_). @leofang

**Reply by leofang** [reply]
Thanks for the investigation, @jessewiles. But doesn't this mean we're circling back to [the beginning of this thread](https://github.com/conda/infrastructure/discussions/649#discussioncomment-4489755), where @beckermr already tried `replace=True` for the copy operation? Is there anything that I misunderstood?

**Reply by jessewiles** [reply]
I _think_ that at the beginning of the thread, the source record was incorrect. In this case, we've corrected the source record and now we need to use the `--replace` flag when copying the source record to update the target record.

**Reply by jakirkham** [reply]
Do we want just `--replace` or do we want `--update` too?

FWIW put together a draft PR ( https://github.com/conda-forge/conda-forge-webservices/pull/558 ) regarding this

Edit: Replaced with PR ( https://github.com/conda-forge/conda-forge-webservices/pull/559 ), which is ready to review

**Reply by jessewiles** [reply]
@jakirkham `--update` will result in a patch in which `null` or missing values in the target record are updated with the value from the source record.  `--replace` will completely replace the target record with the source record. In this case, the package record has a `null` license, so you could probably get away with an `--update`? you don't need both though. `--replace` is more dependable unless you have a good reason to only partially update the target data.

**Reply by jakirkham** [reply]
Thanks Jesse! 🙏

Let's go with `replace` then. Would rather make sure metadata is getting updated when new packages go up (even and especially if that means overwriting something)

This is now in PR ( https://github.com/conda-forge/conda-forge-webservices/pull/559 )

**Reply by beckermr** [reply]
With this flag, would attempting to upload a duplicate package end up overwriting the existing package itself? Conda-forge packages are strictly immutable except for repodata patches.

**Reply by jakirkham** [reply]
What we learned in the other thread is `replace` does more than replace metadata, it will overwrite the package itself, which is too much for our use case. So we are thinking `update` is a better choice for now

That said, it would be nice if there was a way to update metadata that is non-`null` (without overwriting the package). Not sure what that would be called, `--update-all` ?

**Reply by leofang** [reply]
Update: Thanks for help from @jakirkham @beckermr, we merged the above PR to use `update` on copy, and it does [fix the missing license for us](https://github.com/conda-forge/conda-forge-webservices/pull/559#issuecomment-1595561058). 

However, as both Matt and John noted above and elsewhere, this patch is only applicable to missing metadata, not for overwriting/correcting existing (wrong or outdated) metadata. So, I think John's suggestion 
> That said, it would be nice if there was a way to update metadata that is non-`null` (without overwriting the package). Not sure what that would be called, `--update-all`?

is still needed in case of such events, which does happen (ex: https://github.com/conda/infrastructure/discussions/739, where the license hyperlink needs to be fixed).
